runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,2080 @@
+			new_flags: post_flags - pre_flags,
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Called by the initalizer to initialize the disputes module.
+	pub(crate) fn initializer_initialize(now: T::BlockNumber) -> Weight {
+		let config = <configuration::Module<T>>::config();
+
+		let mut weight = 0;
+		for (session_index, candidate_hash, mut dispute) in <Disputes<T>>::iter() {
+			weight += T::DbWeight::get().reads_writes(1, 0);
+
+			if dispute.concluded_at.is_none()
+				&& dispute.start + config.dispute_conclusion_by_time_out_period < now
+			{
+				Self::deposit_event(Event::DisputeTimedOut(candidate_hash));
+
+				dispute.concluded_at = Some(now);
+				<Disputes<T>>::insert(session_index, candidate_hash, &dispute);
+
+				if <Included<T>>::contains_key(&session_index, &candidate_hash) {
+					// Local disputes don't count towards spam.
+
+					weight += T::DbWeight::get().reads_writes(1, 1);
+					continue;
+				}
+
+				// mildly punish all validators involved. they've failed to make
+				// data available to others, so this is most likely spam.
+				SpamSlots::<T>::mutate(session_index, |spam_slots| {
+					let spam_slots = match spam_slots {
+						Some(ref mut s) => s,
+						None => return,
+					};
+
+					// also reduce spam slots for all validators involved, if the dispute was unconfirmed.
+					// this does open us up to more spam, but only for validators who are willing
+					// to be punished more.
+					//
+					// it would be unexpected for any change here to occur when the dispute has not concluded
+					// in time, as a dispute guaranteed to have at least one honest participant should
+					// conclude quickly.
+					let participating = decrement_spam(spam_slots, &dispute);
+
+					// Slight punishment as these validators have failed to make data available to
+					// others in a timely manner.
+					T::PunishValidators::punish_inconclusive(
+						session_index,
+						participating.iter_ones().map(|i| ValidatorIndex(i as _)),
+					);
+				});
+
+				weight += T::DbWeight::get().reads_writes(2, 2);
+			}
+		}
+
+		weight
+	}
+
+	/// Called by the initalizer to finalize the disputes module.
+	pub(crate) fn initializer_finalize() { }
+
+	/// Called by the initalizer to note a new session in the disputes module.
+	pub(crate) fn initializer_on_new_session(notification: &SessionChangeNotification<T::BlockNumber>) {
+		let config = <configuration::Pallet<T>>::config();
+
+		if notification.session_index <= config.dispute_period + 1 {
+			return
+		}
+
+		let pruning_target = notification.session_index - config.dispute_period - 1;
+
+		LastPrunedSession::<T>::mutate(|last_pruned| {
+			let to_prune = if let Some(last_pruned) = last_pruned {
+				*last_pruned + 1 ..= pruning_target
+			} else {
+				pruning_target ..= pruning_target
+			};
+
+			for to_prune in to_prune {
+				<Disputes<T>>::remove_prefix(to_prune, None);
+				<Included<T>>::remove_prefix(to_prune, None);
+				SpamSlots::<T>::remove(to_prune);
```

Comment:
This kind of stuff is very expensive and needs to be limited.

dont take the `None` option here for granted, you should provide an upper bound to how many items can be deleted in one go, and then provide methods for users to continue to clean up the data.

Reply:
What do you think would be a sensible upper bound to do per block initialization? I don't think we should defer this to users without introducing an economic incentive for them to do so

Reply:
We spoke off the thread, but https://github.com/paritytech/polkadot/issues/3469 will address this in near future. For now it's fine.